keras/applications/imagenet_utils.py

```diff
@@ -35,26 +45,27 @@ def preprocess_input(x, data_format=None, mode='caffe'):
         x *= 2.
         return x
 
+    imagenet_mean = [103.939, 116.779, 123.68]
     if data_format == 'channels_first':
         if x.ndim == 3:
             # 'RGB'->'BGR'
             x = x[::-1, ...]
             # Zero-center by mean pixel
-            x[0, :, :] -= 103.939
-            x[1, :, :] -= 116.779
-            x[2, :, :] -= 123.68
+            if _IMAGENET_MEAN_CF is None:
+                _IMAGENET_MEAN_CF = np.array(imagenet_mean, dtype=K.floatx()).reshape((3, 1, 1))
+            x -= _IMAGENET_MEAN_CF
         else:
             x = x[:, ::-1, ...]
-            x[:, 0, :, :] -= 103.939
-            x[:, 1, :, :] -= 116.779
-            x[:, 2, :, :] -= 123.68
+            if _IMAGENET_MEAN_CF_4D is None:
+                _IMAGENET_MEAN_CF_4D = np.array(imagenet_mean, dtype=K.floatx()).reshape((1, 3, 1, 1))
+            x -= _IMAGENET_MEAN_CF_4D
     else:
         # 'RGB'->'BGR'
         x = x[..., ::-1]
         # Zero-center by mean pixel
-        x[..., 0] -= 103.939
-        x[..., 1] -= 116.779
-        x[..., 2] -= 123.68
+        if _IMAGENET_MEAN_CL is None:
+            _IMAGENET_MEAN_CL = np.array(imagenet_mean, dtype=K.floatx()).reshape((1, 1, 3))
+        x -= _IMAGENET_MEAN_CL
```

Comment:
This is missing the 4D case.

Reply:
Also, instead cast to the dtype of `x`. I think `K.dtype(x)` should work with np arrays, but please check.

Reply:
Actually, there's a problem here. The global vars should be able to handle `x` of any `dtype`, since they are global and only set once (use case: if some of our `x`s are fp32 and some are fp64, we don't want to run into type errors). So let's cast to `K.floatx()` as you do, but in case `x` has a different dtype, let's do on-the-fly casting (`x -= _IMAGENET_MEAN_CL.astype(K.dtype(x))` (assuming `K.dtype` does work here).

Reply:
Maybe the global constant should be of type `np.float64` then? Otherwise, if `K.floatx()` changes, that thread-safety issue you mentioned will arise.  Then, maybe `x -= ...` can always (?) just take `np.float64` right-hand-side?